File: frontend/lib/api/managerColumns.ts
/**
 * Formato em COLUNAS da resposta do Manager (migration 161).
 *
 * POR QUE
 * -------
 * A resposta tinha um objeto por linha, com os mesmos ~50 nomes de campo repetidos
 * em todas. Com uma lista por campo, a mesma resposta (10 mil linhas reais, medido
 * em 16/09) cai de 1,98 MB para 1,08 MB na rede e de 16 para 9 MB gravada no
 * IndexedDB. O navegador NÃO ganha lendo colunas — o Chrome lê linhas muito bem —,
 * então a regra é: colunas para trafegar e para gravar; linhas para usar. Este
 * arquivo é o único ponto que converte, nos dois sentidos.
 *
 * O FORMATO (igual ao do backend, `app/services/manager_columns.py`)
 * ------------------------------------------------------------------
 *   { data_columns: [ { campo: valores[] }, ... ], row_count: N, ...topo }
 * `data_columns` é uma lista de blocos porque o Postgres limita 50 campos por
 * `json_build_object`. A linha i é a união de `bloco[campo][i]` de todos os blocos.
 *
 * `row_order[i]` é a POSIÇÃO final da linha i. O banco entrega as listas numa ordem
 * qualquer — ordenar as linhas largas lá custava 75–80 MB em disco por requisição
 * (medido, 16/09) — e este leitor as põe na ordem. Sem `row_order`, a ordem das
 * listas já é a final (é o que a gravação no IndexedDB produz).
 */

export interface ColumnarPayload {
  data_columns: Array<Record<string, unknown[] | null>>
  row_count: number
  row_order?: number[] | null
  [key: string]: unknown
}

export class ColumnarPayloadError extends Error {}

export function isColumnarPayload(value: unknown): value is ColumnarPayload {
  return (
    typeof value === 'object'
    && value !== null
    && 'data_columns' in value
    && 'row_count' in value
  )
}

/**
 * Colunas → linhas. Falha ALTO em coluna desalinhada: um campo deslocado casaria o
 * gasto de um anúncio com o nome de outro, sem erro nenhum na tela.
 */
export function rowsFromColumns(payload: ColumnarPayload): Array<Record<string, unknown>> {
  const n = payload.row_count
  if (!Number.isInteger(n) || n < 0) {
    throw new ColumnarPayloadError(`row_count inválido: ${String(n)}`)
  }
  if (n === 0) return []
  const blocos = payload.data_columns
  if (!Array.isArray(blocos) || blocos.length === 0) {
    throw new ColumnarPayloadError('data_columns ausente com row_count > 0')
  }

  const nomes: string[] = []
  const colunas: unknown[][] = []
  const vistos = new Set<string>()
  for (const bloco of blocos) {
    if (typeof bloco !== 'object' || bloco === null) {
      throw new ColumnarPayloadError('bloco de data_columns não é objeto')
    }
    for (const nome of Object.keys(bloco)) {
      const valores = bloco[nome]
      if (vistos.has(nome)) throw new ColumnarPayloadError(`campo repetido entre blocos: ${nome}`)
      // Atribuir `__proto__` trocaria o protótipo da linha em vez de gravar o valor.
      if (nome === '__proto__') throw new ColumnarPayloadError('campo __proto__ não é aceito')
      if (!Array.isArray(valores) || valores.length !== n) {
        const tam = Array.isArray(valores) ? valores.length : typeof valores
        throw new ColumnarPayloadError(`coluna ${nome} com ${tam} valores, esperado ${n}`)
      }
      vistos.add(nome)
      nomes.push(nome)
      colunas.push(valores)
    }
  }

  // COMO MONTAR IMPORTA (medido em 16/09, Node 22 = motor do Chrome, 10 mil linhas):
  //   JSON em linhas (o formato antigo) ........ 22,9 MB retidos, 53–64 ms
  //   objeto vazio + campo a campo (o óbvio) ... 50,4 MB  <- mais que o dobro
  //   cópia de um MOLDE criado pelo JSON.parse . 22,8 MB, 57–59 ms  <- este
  //   `new Function` com os campos fixos ....... 26,8 MB, e a CSP de produção não
  //                                              permite 'unsafe-eval'
  // O molde nasce do JSON.parse com todos os campos, então as cópias já têm a forma
  // final e a atribuição só troca valores — sem o objeto crescer campo a campo.
  const ordem = payload.row_order
  let indices: number[] | null = null
  if (ordem != null) {
    if (!Array.isArray(ordem) || ordem.length !== n) {
      throw new ColumnarPayloadError(`row_order com tamanho errado (esperado ${n})`)
    }
    indices = Array.from({ length: n }, (_, i) => i)
    indices.sort((a, b) => (ordem[a] as number) - (ordem[b] as number))
  }

  const k = nomes.length
  const molde = JSON.parse(`{${nomes.map((nome) => `${JSON.stringify(nome)}:null`).join(',')}}`) as Record<string, unknown>
  const rows = new Array<Record<string, unknown>>(n)
  for (let pos = 0; pos < n; pos++) {
    const i = indices === null ? pos : indices[pos]
    const row = { ...molde }
    for (let j = 0; j < k; j++) row[nomes[j]] = colunas[j][i]
    rows[pos] = row
  }
  return rows
}

/**
 * A resposta no formato que o resto do app sempre usou (`data` = linhas). Aceita os
 * dois formatos: a resposta sem packs e a de uma rota antiga já vêm em linhas.
 */
export function asRowPayload<T extends Record<string, unknown>>(payload: T | ColumnarPayload): T {
  if (!isColumnarPayload(payload)) return payload as T
  const { data_columns: _cols, row_count: _n, row_order: _ord, ...topo } = payload
  void _cols
  void _n
  void _ord
  return { ...topo, data: rowsFromColumns(payload) } as unknown as T
}

/**
 * Linhas → colunas, para GRAVAR (IndexedDB). Só empacota quando todas as linhas têm
 * exatamente os mesmos campos: senão, "campo ausente" viraria "campo nulo" na volta
 * — e o app distingue os dois em alguns lugares. Nesse caso devolve `null` e quem
 * chamou grava as linhas como sempre.
 */
export function columnsFromRows(rows: ReadonlyArray<Record<string, unknown>>): Pick<ColumnarPayload, 'data_columns' | 'row_count'> | null {
  const n = rows.length
  if (n === 0) return { data_columns: [], row_count: 0 }
  const nomes = Object.keys(rows[0])
  const k = nomes.length
  const colunas: unknown[][] = nomes.map(() => new Array(n))
  for (let i = 0; i < n; i++) {
    const row = rows[i]
    if (Object.keys(row).length !== k) return null
    for (let j = 0; j < k; j++) {
      const nome = nomes[j]
      if (!(nome in row)) return null
      const v = row[nome]
      // `undefined` não existe em JSON: viraria null e mudaria a linha na volta.
      if (v === undefined) return null
      colunas[j][i] = v
    }
  }
  const bloco: Record<string, unknown[]> = {}
  nomes.forEach((nome, j) => { bloco[nome] = colunas[j] })
  return { data_columns: [bloco], row_count: n }
}

File: backend/app/services/manager_columns.py
"""
Leitor do formato em COLUNAS das `fetch_manager_rankings_v161/v162` (migrations 161 e 162).

POR QUE COLUNAS
---------------
A resposta do Manager tem um objeto por linha, com os mesmos ~50 nomes de campo
repetidos em todas elas. Medido em 16/09 com 10 mil linhas reais: com uma lista
por campo, a resposta cai de 1,98 MB para 1,08 MB na rede (os valores parecidos
ficam juntos e comprimem melhor). O navegador volta a montar as linhas num ponto
só — ele lê linhas muito bem; colunas servem para trafegar e para gravar.

O FORMATO
---------
    {
      "data_columns": [ {"group_key": [...], "spend": [...], ...}, {...}, ... ],
      "row_count": N,
      ...demais campos de topo iguais aos de antes (averages, names, pagination...)
    }

`data_columns` é uma lista de BLOCOS porque o `json_build_object` do Postgres
aceita no máximo 100 argumentos (50 campos). Cada bloco tem listas de tamanho N,
e a linha i é a união de `bloco[campo][i]` de todos os blocos.

`row_order` (lista de tamanho N): a POSIÇÃO final da linha i. As listas saem do
banco numa ordem qualquer — ordenar as linhas largas lá custava 75–80 MB em disco
por requisição (medido, 16/09) — e o leitor as põe na ordem aqui. Sem `row_order`,
a ordem das listas já é a final (é o que a gravação no navegador produz).

EM PEDAÇOS (162)
----------------
A v162 entrega a MESMA resposta como uma lista de objetos, sem o envelope:

    [ {"row_count": N, "row_order": [...], "names": ..., ...}, {"group_key": [...]},
      {"spend": [...]}, ... ]

Montar o envelope dentro do banco custava ~11x o tamanho da resposta em memória
(medido em 16/09: 957 MB de pico com 55 MB de resposta, numa máquina de ~950 MB);
em pedaços, cada um vai para o disco assim que fica pronto. O pedaço de metadados é o
que tem `row_count`; os demais são blocos de colunas. A ordem dos pedaços não importa.
`from_parts` devolve o formato de cima, e o resto do leitor não muda.

O leitor do navegador (`frontend/lib/api/managerColumns.ts`) faz exatamente isto;
os dois são conferidos contra o mesmo fixture.
"""
from __future__ import annotations

from typing import Any, Dict, List


class ColumnarPayloadError(ValueError):
    """A resposta em colunas veio malformada — melhor falhar alto do que mostrar
    linhas trocadas (um campo deslocado casaria o gasto de um anúncio com o nome
    de outro, sem erro nenhum)."""


def from_parts(parts: List[Any]) -> Dict[str, Any]:
    """Resposta em pedaços (v162) -> o formato com `data_columns` (v161)."""
    metas = [p for p in parts if isinstance(p, dict) and "row_count" in p]
    if len(metas) != 1:
        raise ColumnarPayloadError(f"esperado 1 pedaço de metadados, vieram {len(metas)}")
    blocos = []
    for p in parts:
        if p is metas[0]:
            continue
        if not isinstance(p, dict):
            raise ColumnarPayloadError("pedaço da resposta não é objeto")
        blocos.append(p)
    out = dict(metas[0])
    if "data_columns" in out:
        raise ColumnarPayloadError("pedaço de metadados com data_columns")
    out["data_columns"] = blocos
    return out


def rows_from_columns(payload: Dict[str, Any]) -> List[Dict[str, Any]]:
    n = payload.get("row_count")
    if not isinstance(n, int) or n < 0:
        raise ColumnarPayloadError(f"row_count inválido: {n!r}")
    if n == 0:
        return []
    blocos = payload.get("data_columns")
    if not isinstance(blocos, list) or not blocos:
        raise ColumnarPayloadError("data_columns ausente com row_count > 0")

    campos: List[tuple] = []
    vistos: set = set()
    for bloco in blocos:
        if not isinstance(bloco, dict):
            raise ColumnarPayloadError("bloco de data_columns não é objeto")
        for nome, valores in bloco.items():
            if nome in vistos:
                raise ColumnarPayloadError(f"campo repetido entre blocos: {nome}")
            if not isinstance(valores, list) or len(valores) != n:
                tam = len(valores) if isinstance(valores, list) else type(valores).__name__
                raise ColumnarPayloadError(f"coluna {nome} com {tam} valores, esperado {n}")
            vistos.add(nome)
            campos.append((nome, valores))

    ordem = payload.get("row_order")
    if ordem is None:
        indices = range(n)
    else:
        if not isinstance(ordem, list) or len(ordem) != n:
            raise ColumnarPayloadError(f"row_order com tamanho errado (esperado {n})")
        indices = sorted(range(n), key=ordem.__getitem__)
    return [{nome: valores[i] for nome, valores in campos} for i in indices]


def as_row_payload(payload: Any) -> Dict[str, Any]:
    """A resposta em colunas (inteira ou em pedaços) no formato ANTIGO (`data` =
    lista de linhas).

    Para consumidores internos do backend que leem linhas (filhos de campanha) e
    para os diferenciais da 161/162.
    """
    if isinstance(payload, list):
        payload = from_parts(payload)
    if "data_columns" not in payload:
        return payload
    out = {k: v for k, v in payload.items() if k not in ("data_columns", "row_count", "row_order")}
    out["data"] = rows_from_columns(payload)
    return out
